modules/spellfiles.nix

```diff
@@ -0,0 +1,175 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  inherit (lib) types;
+
+  spellfilesHashes = import ../generated/spellfiles.nix;
+
+  availableSpellfiles = lib.attrNames spellfilesHashes;
+
+  # en.utf-8.spl -> en
+  extractLanguage = filename: builtins.head (builtins.split "\\." filename);
+
+  availableLanguages = lib.pipe availableSpellfiles [
+    (map extractLanguage)
+    lib.unique
+  ];
+
+  cfg = config.spellfiles;
+in
+{
+  options.spellfiles = {
+    enable = lib.mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to automatically fetch and install vim spellfiles (`.spl` and `.sug`).
+
+        If enabled, Nixvim will automatically populate `spell/en.utf-8.spl`-type files by fetching
+        them from [the official vim FTP server](https://ftp.nluug.nl/pub/vim/runtime/spell/).
+
+        This module allows choosing which spellfiles to install.
+      '';
+    };
+
+    languages = lib.mkOption {
+      type = with types; either (enum [ "auto" ]) (listOf (enum availableLanguages));
+      default = "auto";
+      example = [
+        "en"
+        "fr"
+      ];
+      description = ''
+        By default, spellfiles from the languages listed in `opts.spelllang` will be fetched (or
+        `"en"` if `opts.spelllang` is not defined).
+
+        Set to `[ ]` to disable this behavior.
+      '';
+    };
+
+    includeSuggestions = lib.mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to also fetch suggestion files (`.sug`) when available.
+
+        Warning, using suggestion files can lead to higher memory usage.
+
+        See https://neovim.io/doc/user/spell.html#spell-NOSUGFILE.
+      '';
+    };
+
+    spellfiles = lib.mkOption {
+      type = with types; listOf (enum availableSpellfiles);
+      default = [ ];
+      description = ''
+        Explicitly specify spellfiles to install.
+
+        Possible values correspond to the spellfiles available at https://ftp.nluug.nl/pub/vim/runtime/spell/.
+      '';
+      example = [
+        "en.utf-8.spl"
+        "en.utf-8.sug"
+        "fr.utf-8.spl"
+      ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    extraFiles =
+      let
+        /*
+          Infers the list of language ids from `opts.spelllang`:
+
+          Turns: "en_us,fr,de"
+          into: [ "en" "fr" "de" ]
+          # https://neovim.io/doc/user/spell.html#spell-load
+        */
+        languagesFromSpelllang = lib.pipe (config.spelllang or "en") [
+
+          # "en_us,fr,de" -> [ "en_us" "fr" "de" ]
+          (lib.splitString ",")
+
+          # Remove region from language id ("en_us" -> "en")
+          (map (lang: lib.head (builtins.match "^([^_]+)" lang)))
+        ];
+        languages = if cfg.languages == "auto" then languagesFromSpelllang else cfg.languages;
```

Comment:
Fancy idea: could we move this to `cfg.languages` apply function?

That way users can _read_ the value to figure out what languages their "auto" definition actually ended up producing.

Format normalisation could also be done there.

Also makes me wonder if the option should be a list type? 🤔 And maybe auto should be a separate bool option?

Reply:
> I don't really see a benefit there...

Definition merging, and just generally having a more structural config.

If the option is a list-type, then I can define `[ "en" ]` in one module, define `[ "fr" ]` in another module, and it will be merged as `[ "en" "fr" ]`. With a str-type option, merging is not supported.

Reply:
Or maybe you meant the separate auto-option? I think the benefit of that separation is simplifying the types and the impl, making things more obvious to end-users.

If you had a separate auto-toggle, you wouldn't need any `if value == "auto"`; in fact, you wouldn't need an `apply` at all: you could just have an extra definition that sets `config.spellfiles.languages = lib.pipe ...`